### ISSUE TITLE
cowboy_clock crashes during lookup of rfc1123

### ISSUE THREAD

**Issue by iguberman** [ninenines/cowboy#1228]
I get below error log intermittently, possibly during high load in the system.  

```2017-10-20 04:56:50.535 [error] <0.4194.0>@cowboy_clock:rfc1123:54 CRASH REPORT Process <0.4194.0> with 0 neighbours crashed with reason: bad argument in call to ets:lookup_element(cowboy_clock, rfc1123, 2) in cowboy_clock:rfc1123/0 line 54
2017-10-20 04:56:50.535 [error] <0.720.0> Lager event handler error_logger_lager_h exited with reason {'EXIT',{{case_clause,[http,<0.4193.0>,1,<0.4194.0>,badarg,[{ets,lookup_element,[cowboy_clock,rfc1123,2],[]},{cowboy_clock,rfc1123,0,[{file,"/drone/src/code.xaptum.xyz/xaptum-development/xmx/_build/default/lib/cowboy/src/cowboy_clock.erl"},{line,54}]},{cowboy_req,response_headers,2,[{file,"/drone/src/code.xaptum.xyz/xaptum-development/xmx/_build/default/lib/cowboy/src/cowboy_req.erl"},{line,778}]},{cowboy_req,do_reply,4,[{file,"/drone/src/code.xaptum.xyz/xaptum-development/xmx/_build/defau..."},...]},...]]},...}}
2017-10-20 04:56:50.588 [error] <0.4197.0>@cowboy_clock:rfc1123:54 CRASH REPORT Process <0.4197.0> with 0 neighbours crashed with reason: bad argument in call to ets:lookup_element(cowboy_clock, rfc1123, 2) in cowboy_clock:rfc1123/0 line 54
2017-10-20 04:56:50.588 [error] <0.720.0> Lager event handler error_logger_lager_h exited with reason {'EXIT',{{case_clause,[http,<0.4196.0>,1,<0.4197.0>,badarg,[{ets,lookup_element,[cowboy_clock,rfc1123,2],[]},{cowboy_clock,rfc1123,0,[{file,"/drone/src/code.xaptum.xyz/xaptum-development/xmx/_build/default/lib/cowboy/src/cowboy_clock.erl"},{line,54}]},{cowboy_req,response_headers,2,[{file,"/drone/src/code.xaptum.xyz/xaptum-development/xmx/_build/default/lib/cowboy/src/cowboy_req.erl"},{line,778}]},{cowboy_req,do_reply,4,[{file,"/drone/src/code.xaptum.xyz/xaptum-development/xmx/_build/defau..."},...]},...]]},...}}```

**Comment by essen** [comment]
Is there anything else? Because I think this could happen for one of two cases:

- the cowboy_clock process has crashed and is being restarted; the ets table is temporarily gone
- there is a bug in ets and the writes are not atomic

Would be good if you can confirm whether cowboy_clock restarted (I guess the pid number can give a good hint) and if it has, see if you can find the crash reason.

**Comment by alekras** [comment]
Hi Irina,
I have the same issue when I did not start cowboy as application. The cowboy_clock process is started as worker by cowboy_sup that is started by cowboy_app.
I hope it will help in your debugging.
Regards,
Alexei.

**Comment by iguberman** [comment]
Hi Alexei,

Thanks so much!

I have cowboy app running for sure as it's working for me all along.
However, I saw this error again and this time I was mindful of what you
said above and realized that this happened while my server was being
restarted (obviously along with all the other apps restarting with it -- so
looks like something is still trying to use cowboy when it's already
stopped/stopping and that's why I get this error).   I can't be sure now
about the other times I was seeing it, maybe it was always this situation.
I'll keep an eye on it to see if I can reproduce when the server isn't
restarting, but I suspect now that this was probably always the case.

Cheers,
Irina.

**Comment by essen** [comment]
I can see that happening if Ranch supervises the connections because those will still run while Cowboy is being restarted, and if Cowboy is not there then cowboy_clock isn't either, so the ets table doesn't exist (anymore).

I suppose a graceful way to deal with that is to catch the error and then compute the value directly if the ets table is gone.

**Comment by iguberman** [comment]
Hi Loïc,

I've been watching our logs and seems like it does happen upon server exit.
  So I guess it isn't really an issue.

Thanks so much for your help,
Irina.

**Comment by essen** [comment]
Cheers, good to know! I will still do the proposed change in my previous comment so that these logs don't appear anymore and to not have connections get killed because of this.

**Comment by essen** [comment]
Fixed locally, this will be part of 2.1. Thanks!

**Comment by iguberman** [comment]
Awesome!  Thanks so much Loïc!
